torchmdnet/datasets/custom.py

```diff
@@ -30,87 +60,160 @@ class Custom(Dataset):
+            f"does not match number of embed files {len(self.files['z'])}."
         )
         if self.has_energies:
-            assert len(self.coordfiles) == len(self.energyfiles), (
-                f"Number of coordinate files {len(self.coordfiles)} "
-                f"does not match number of energy files {len(self.energyfiles)}."
+            self.files["y"] = sorted(glob.glob(energyglob))
+            self.fields.append(("y", "energy", torch.float32))
+            assert len(self.files["pos"]) == len(self.files["y"]), (
+                f"Number of coordinate files {len(self.files['pos'])} "
+                f"does not match number of energy files {len(self.files['y'])}."
             )
         if self.has_forces:
-            assert len(self.coordfiles) == len(self.forcefiles), (
-                f"Number of coordinate files {len(self.coordfiles)} "
-                f"does not match number of force files {len(self.forcefiles)}."
+            self.files["neg_dy"] = sorted(glob.glob(forceglob))
+            self.fields.append(("neg_dy", "forces", torch.float32))
+            assert len(self.files["pos"]) == len(self.files["neg_dy"]), (
+                f"Number of coordinate files {len(self.files['pos'])} "
+                f"does not match number of force files {len(self.files['neg_dy'])}."
             )
-
-        print("Number of files: ", len(self.coordfiles))
-
+        print("Number of files: ", len(self.files["pos"]))
+        self.cached = False
+        if read_as_hdf5 is not None:
+            hdf5_file = read_as_hdf5
+            write_as_hdf5(
+                self.files,
+                hdf5_file,
+            )
+            self.hdf5_dataset = HDF5(hdf5_file)
+        total_data_size = self._initialize_index()
+        print("Combined dataset size {}".format(len(self.index)))
+        # If the dataset is small enough, load it whole into CPU memory
+        data_size_limit = preload_memory_limit * 1024 * 1024
+        if total_data_size < data_size_limit:
+            self.cached = True
+            print(
+                "Preloading Custom dataset (of size {:.2f} MB)".format(
+                    total_data_size / 1024**2
+                )
+            )
+            self._preload_data()
+        else:
+            self._store_numpy_memmaps()
+
+    def _preload_data(self):
+        """Load the input files as Torch tensors.
+        Each file can have different number of atoms, so each one is stored in a different tensor.
+        """
+        self.stored_data = {}
+        self.stored_data["pos"] = [
+            torch.from_numpy(np.load(f)) for f in self.files["pos"]
+        ]
+        self.stored_data["z"] = [
+            torch.from_numpy(np.load(f).astype(int))
+            .unsqueeze(0)
+            .expand(self.stored_data["pos"][i].shape[0], -1)
+            for i, f in enumerate(self.files["z"])
+        ]
+        if self.has_energies:
+            self.stored_data["y"] = [
```

Comment:
Is `torch.from_numpy()` what you really need? Take into account that if you modify the tensor then these changes will be reflected in the ndarray and vice versa. Moreover, the returned tensor is not resizable. I would say `torch.tensor()` is fine.  I think you can use also here the `mmap_mode="r"`, it will work only with `torch.tensor()`

Reply:
But I do not want to mmap these ones, I want to load them into memory, they will never be resized or mutated in any way, only accesed.